RestInterface/cafRestSession.cpp

```diff
@@ -0,0 +1,513 @@
+    for ( size_t i = 1; i < targetComponents.size(); ++i )
+    {
+        params.push_back( targetComponents[i] );
+    }
+
+    std::shared_ptr<caffa::rpc::RestServiceInterface> service;
+
+    auto pathComponents = caffa::StringTools::split<std::list<std::string>>( path, "/", true );
+
+    if ( pathComponents.size() >= 1u )
+    {
+        auto docOrServiceComponent = pathComponents.front();
+        service                    = findRestService( docOrServiceComponent, services );
+        if ( service )
+        {
+            pathComponents.pop_front();
+        }
+    }
+
+    if ( !service )
+    {
+        service = findRestService( "object", services );
+    }
+
+    CAFFA_ASSERT( service );
+
+    if ( service->requiresAuthentication( pathComponents ) )
+    {
+        auto authorisation = req[http::field::authorization];
+        auto trimmed       = caffa::StringTools::replace( std::string( authorisation ), "Basic ", "" );
+
+        if ( !authenticator->authenticate( decodeBase64( trimmed ) ) )
+        {
+            if ( authorisation.empty() )
+            {
+                send( createResponse( http::status::unauthorized, "Need to provide password" ) );
+                return nullptr;
+            }
+
+            send( createResponse( http::status::forbidden, "Failed to authenticate" ) );
+            return nullptr;
+        }
+    }
+
+    nlohmann::json jsonArguments = nlohmann::json::object();
+    if ( !req.body().empty() )
+    {
+        try
+        {
+            jsonArguments = nlohmann::json::parse( req.body() );
+        }
+        catch ( const nlohmann::detail::parse_error& )
+        {
+            CAFFA_ERROR( "Could not parse arguments \'" << req.body() << "\'" );
+            send( createResponse( http::status::bad_request,
+                                  std::string( "Could not parse arguments \'" ) + req.body() + "\'" ) );
+            return nullptr;
+        }
+    }
+    nlohmann::json jsonMetaData = nlohmann::json::object();
+    for ( auto param : params )
+    {
+        auto keyValue = caffa::StringTools::split<std::vector<std::string>>( param, "=", true );
+        if ( keyValue.size() == 2 )
+        {
+            if ( auto intValue = caffa::StringTools::toInt64( keyValue[1] ); intValue )
+            {
+                jsonMetaData[keyValue[0]] = *intValue;
+            }
+            else if ( auto doubleValue = caffa::StringTools::toDouble( keyValue[1] ); doubleValue )
+            {
+                jsonMetaData[keyValue[0]] = *doubleValue;
+            }
+            else if ( strcasecmp( "true", keyValue[1].c_str() ) == 0 )
+            {
+                jsonMetaData[keyValue[0]] = true;
+            }
+            else if ( strcasecmp( "false", keyValue[1].c_str() ) == 0 )
+            {
+                jsonMetaData[keyValue[0]] = false;
+            }
+            else
+            {
+                jsonMetaData[keyValue[0]] = keyValue[1];
+            }
+        }
+    }
```

Comment:
I think repeated keys are allowed in the parameters (weirdly), e.g `?key=1&key=2`

Here the first value set to key will be overwritten. Might not be a problem for us, though.

Reply:
I think we'll accept that.